### PR TITLE
Add library refcounting to allow for library cleanup without using atexit()

### PR DESCRIPTION
**Pull request by nhorman**
With the removal of atexit driven cleanup, we potentially leak
    referenced memory.  While this is not strictly an error, as the memory
    is reachable, it would be nice to give software components a method for
    handling proper cleanup of libcrypto so as to avoid that where possible,
    particularly in those use cases where subordonate dsos might use
    openssl without the top level application being aware of this
    
To do this, lets introduce two new API calls to the libcrypto public
    api:
    OPENSSL_add_library_user()
    OPENSSL_cleanup_ex()
    
OPENSSL_add_library_user() increments the global library reference
    count, while OPENSSL_cleanup_ex() decrements it, freeing the global
    library resources only when the reference count reaches zero (In the
    same way OPENSSL_cleanup() did unilaterally previously).
    
In this commit we also deprecate OPENSSL_cleanup(), so as to warn users
    about its future use/lack thereof, and change its behavior slightly.
    With this commit, OPENSSL_cleanup() is now a no-op if any other software
    component has incremented the reference count of the library.  IFF no
    other component has participated in library reference counting does
    OPENSSL_cleanup() free library resources, providing an amount of
    backward compatibility.
    
Update the OpenSSL docs, particularly the migration guide to flag this
    change in behavior/api, so that developers undertaking migration are
    aware of this new reference counting requirement.
    
Note: If a sofware component uses libcrypto but fails to participate in
    this reference counting model, it may occur that usage in that component
    will run afoul of a use-after-free condition as they do not hold a
    reference to the library.  this is considered a programming error and
    needs to be fixed in that software component.  However, on platforms that support automated library load/unload events (like via constructors/destructors, or DllMain events), this process can be done automatically, avoiding this corner case.

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: doc/man7/ossl-guide-migration.pod [hunk 1]
```diff
@@ -42,6 +42,32 @@ Memory leak detectors may report there is allocated, but still reachable,
 allocated memory at application exit. If clean report is desired, then
 application must call OPENSSL_cleanup() explicitly before main() returns.
 
+=head3 Introduction of library reference counting
+
+OpenSSL 3.x encountered various issues with automatic library cleanup via
+atexit() registration.  To address these issues, OpenSSL is moving to a library
+reference counting implementation.  Users of the OpenSSL 4.0 libcrypto library
```
Comment (bob-beck) [inline doc/man7/ossl-guide-migration.pod:49]: I think we need a little more clarity, which may affect the naming here. 

what's a "User" of the library and who needs to call this? 

1) Is a standalone main program that uses libcrypto, no other libraries, and exits a "user"
2) Is another statically linked library that expects to call libcrypto functions a user? 
3) Is another dynamically linked library linked to an executable at compiled time that expects to call libcrcrypto functions a user? 
4) Is another dynamically linked object that users libcrypto functions which is going to be dlloaded into a main program until exit a user? (ok that's basically 3)
5) Is another dynamically linked object that uses libcrypto functions that is going to be dlloaded and dlunloaded a user?

File: doc/man7/ossl-guide-migration.pod [hunk 2]
```diff
@@ -42,6 +42,32 @@ Memory leak detectors may report there is allocated, but still reachable,
...
+
+Note that OPENSSL_cleanup() has been deprecated in OpenSSL 4.0.0.  While this
+API call may still be used, its behavior is altered in light of the new
+reference counted API above.  If no other software component within a process
+uses the OPENSSL_add_library_user() and OPENSSL_cleanup_ex() calls above,
+OPENSSL_cleanup() functions exactly as it does in prior releases.  However, if
+any other software component makes a call to OPENSSL_add_library_user() or
+OPENSSL_cleanup_ex(), then any calls to OPENSSL_cleanup become no-ops.  This is
+to ensure that the library does not free resources prematurely.  Note that the
+result of this design change is that library memory resources may leak on
```
Comment (bob-beck) [inline doc/man7/ossl-guide-migration.pod:67]: I really hate this language of "leak on application exit" - because really, this is not a leak, it's normal program behaviour. ONLY VALGRIND CARES :) 

Can we come up with some sort of better word?  "memory that is not free'ed until exit?" (I'm not good at this?)

Comment (bob-beck) [reply doc/man7/ossl-guide-migration.pod:67]: "heap allocated globals only being free'd on exit?"

File: doc/man7/ossl-guide-migration.pod (hunk 1 above)
Comment (nhorman) [reply doc/man7/ossl-guide-migration.pod:49]: all of the above, its any software component (application/library/other), that wants to use the openssl libcrypto api safely, which can't rely on another component to avoid deinitalizing libcrypto prior to it being done with its use.  In short a "user" is a user of the libcrypto library api.

The guidance here is, if you are working on a software project, and you link to libcrypto.so, use add_library_user before any other libcrypto call, and only use cleanup_ex after you are sure that no other calls to libcrypto apis will be made.

File: doc/man7/ossl-guide-migration.pod (hunk 2 above)
Comment (nhorman) [reply doc/man7/ossl-guide-migration.pod:67]: I too am not a fan of it, but its the language that users who complain about this behavior will use.  we've had, IMHO too many questions/complaints about memory leaks to come up with softer/more accurate language.  At least if we say "leak", we can say we warned about this behavior if you don't follow this new api guideline.

Comment (davidben) [reply doc/man7/ossl-guide-migration.pod:67]: Are the "leaks" in question still reachable from globals at exit time? All the leak detection tools we use do a reachability check from globals, which seems to avoid most of this sort of fuss.

File: doc/man7/ossl-guide-migration.pod (hunk 1 above)
Comment (bob-beck) [reply doc/man7/ossl-guide-migration.pod:49]: Well, that's kind of yukky, so, basically what you're saying for this is only to appease valgrind, we're going back to
non-automatic init, and everyone needs to call this.

while, yes, if you don't care about valgrind, you won't need to, effectively with linux distros compiling with valgrind and whining about it, people will be forced to re-introduce a new call

and better yet it's different than before.  Why not just ressurect OPENSSL_init_crypto and make it do this?

Comment (bob-beck) [reply doc/man7/ossl-guide-migration.pod:49]: the situation I'd love to avoid in upstream software is forcing people to carry

#if defined OPENSSL_WHATEVER_DETECT_OPENSSL_4.
OpenSSL_add_library_user();
#elif defined OPENSSL_WHATEVER DETECT_OPENSSL_3 || defined (somethignsomething boringssl) || defined (something somthign libressl)
/* do nothing */
#elif defined (OPENSSL_WHATEVER_1_1_TOMFOOLERY)
OPENSL_init_crypto;
.


followed by
#if defined OPENSSL_WHATEVER_DETECT_OPENSSL_4.
OpenSSL_cleanup_ex();
#elif defined OPENSSL_WHATEVER DETECT_OPENSSL_3 || defined (somethignsomething boringssl) || defined (something somthign libressl)
/* do nothing */
#elif defined (OPENSSL_WHATEVER_1_1_TOMFOOLERY)
OPNSSL_cleanup();
#endif

I think if we a determined to undo initialization-less openssl, we should not introduce new stuff for this..

If it's in the existing functions, today, things can just call them everywhere no matter what version.

File: doc/man7/ossl-guide-migration.pod (hunk 2 above)
Comment (bob-beck) [reply doc/man7/ossl-guide-migration.pod:67]: Yes, normal leak detection tools do this ;) 

Why don't we just make OpenSSL refuse to compile in the presence of valgrind :)

Comment (nhorman) [reply doc/man7/ossl-guide-migration.pod:67]: @davidben except the fuss isn't about this being a non-problem, its about users complaining that it is a problem for them for various and sundry reasons.  We get numerous issues about still reachable memory being not freed.  A few notable issues from recent memory:

https://github.com/openssl/openssl/issues/28757 - user complained that prior versions didn't show reachable leaks, but openssl 3.5 did

https://github.com/openssl/openssl/issues/28770 - bind9 uses their own internal memory leak checker, which IMHO is mis-designed, but nevertheless complains about openssl

https://github.com/openssl/openssl/issues/28951 - another "this didn't happen before" complaint

https://github.com/openssl/openssl/issues/27378 - more of the same

I suppose the alternate solution here is simply to remove OPENSSL_cleanup() in 4.0, or make it a no-op, to ensure that no cleanup can be preformed prematurely (though that is going to impact users which load libcrypto.so via dlopen/dlclose repeatedly, which happens a in a suprising number of use cases), and document the fact that we will have reachable allocated memory at program shutdown, but thats not going to quell the complaints about it.  I would much prefer to at least give some sort of public mechanism to do this cleanup.

File: doc/man7/ossl-guide-migration.pod (hunk 1 above)
Comment (nhorman) [reply doc/man7/ossl-guide-migration.pod:49]: OPENSSL_init_crypto isn't a good api to do this in (and never was), because of how it prototyped, and how its used:
```
OPENSSL_init_crypto(uint64_t init_flags, SETTINGS *settings);
```
Its use implies your setting up openssl to initalize with various flags/settings that may not be reflected if the library has already been setup.  Its also lousy, because its used throughout libcrypto in various locations for on-demand initialization, meaning we would have to split the function into a public call and an internal call to keep the reference counting balanced.  And users would still have to modify their code to make use of it.

And yes, users would have to do some ifdeffery to support multiple versions, but we're already asking them to do a significant amount of that anyway, with the removal of engines/SSLv3/etc.

The bottom line is that, if we want to provide users a way to clean up libcrypto, we need a set of bounding functions to define when a software component starts and stops using openssl, which we don't currently have (since we previously just did on demand initialization, and automated cleanup via atexit(), which we've decided was a very bad idea for a slew of other reasons).

the only alternative I see to creating new apis to do this, is to not allow cleanup at all (so as to avoid premature cleanup, and the crashes that result, referencing issues like https://github.com/openssl/openssl/issues/25294), document the fact that we expect reachable memory reports with tools like valgrind that may show up as legitimate leaks in other tools (see https://github.com/openssl/openssl/issues/28770), and have all the "no its not a problem" discussions that result.  I'm happy to go that way if thats the consensus, but I'm personally not a fan.

File: doc/man7/ossl-guide-migration.pod (hunk 2 above)
Comment (davidben) [reply doc/man7/ossl-guide-migration.pod:67]: @bob-beck Well, valgrind is a dynamic analysis tool, so you can't detect it at compile-time. There's probably some way to detect at runtime whether you're in valgrind, but dubiously worth it. :-) Especially since valgrind's defaults are just fine (below).

@nhorman Ah, I guess people are passing some flags to valgrind to make it still report it? Valgrind does do a reachability analysis and even documents this as not a problem:
https://valgrind.org/docs/manual/mc-manual.html#:~:text=%22Still%20reachable%22.%20This,such%20blocks%20individually.

Playing around with valgrind myself, valgrind's documentation is right in that it doesn't show these by default. `--leak-check=full` is not sufficient to trigger this. You have to go even further and add `--show-leak-kinds=all`.

I suspect you're mostly running into a variant of [Hyrum's Law](https://www.hyrumslaw.com/) where, because this overeager check used to work with OpenSSL, folks used it without really evaluating whether it's wrong. And so, like you say, should you all decide to not care about this, I think it should eventually settle with clear documentation, after an initial burst of reports like this.

Personally, that's the path I would go with, as I think this kind of stuff is really not worth the effort to chase down. If you all want to go down that path, 4.0 seems as fine a time as any to do it. (But this is just my 2 cents. Libraries that believe in such cleanup aren't unheard of either.)

Comment (davidben) [reply doc/man7/ossl-guide-migration.pod:67]: If you all do decide to keep supporting this, I'd second @bob-beck's original comment though. These are decidedly not leaks, and you shouldn't document it as such. Then we don't teach the next generation of application developers that this machinery is how you avoid leaks. They're really about avoiding false positives in overeager analysis tools, and it's worth saying as much.

Also, thinking about this some more, this refcounting story seems to have a bit of a problem with it, unless I'm misunderstanding things? It seems to basically force every OpenSSL user to take on this false positive complexity, just in case someone else does:

Suppose your application X uses library Y and library Z, both of which use OpenSSL. X uses correct leak-detection tools and does not care about these false positives. So does Y and does not use these new APIs, because that would require Y expose init/cleanup APIs. That is extra burden for both X and Y, and neither of them want to play this game.

However, Z's developer misunderstood how memory leaks work, perhaps because the documentation falsely called this a memory leak, so they call `OPENSSL_add_library_user` and `OPENSSL_cleanup_ex` throughout their code. It all works fine when Z is used standalone (Z uses it such that it is never actively using OpenSSL when its refcount is zero), but as soon as it is linked in with Y, the refcount flaps to zero (because Y did not bother to call `OPENSSL_add_library_user`, since it knew this whole thing was nonsense) and OpenSSL gets unloaded under Y's feet.

We then end up in a situation where you have to defensively `OPENSSL_add_library_user`, and now this becomes yet another incant you have to use to call into a library that was really just used to do a few hashes.

File: doc/man7/ossl-guide-migration.pod (hunk 1 above)
Comment (bob-beck) [reply doc/man7/ossl-guide-migration.pod:49]: So #28770 is a "custom tool" written by them, I'm not convinced we should care about that. 

Valgrind is really the only issue getting in the way of "no it's not a problem" that I'm aware of that's real .

I think we should consider other options on the realm of "no it's not a problem", and simply not do it, and provide advice to how to run under valgrind.  It's galling to have to have everyone do this for one special
case, for one tool .

1) Can we ship a valgrind suppression file with the library for the things atexit would have freed? then we say
"don't worry" and "run with this suppression file if you are using valgrind"

2) Keep the atexit handler and install it only when valgrind is present or when NDEBUG is not defined.  (I.E. it is installed in debug builds that you can then mess about with valgrind, but just like asserts is it not present in production builds, providing attack surface

File: doc/man7/ossl-guide-migration.pod (hunk 2 above)
Comment (nhorman) [reply doc/man7/ossl-guide-migration.pod:67]: @davidben, yes, some (most) people who  gripe about this run with --leak-check=full --show-leak kinds all, and expect a zero output.  We try to tell them they're false positives, and we get responses that range from "ok, thank you" to "this didn't happen previously (for various and sundry reasons)" to "I don't care, I expect this tool to report no leaks in my configuration".  We can certainly take a "pound sand, its not a problem approach", but we've been doing that for years now (particularly with no-atexit builds), and its an ongoing struggle.

And yes, in the case where application A uses libs x and y, and only one participates in the ref counting, we are at risk from crashing in a use after free scenario, but thats no worse than the situation we're in now (see https://github.com/openssl/openssl/issues/25294 in which at atexit cleanup happened before a library destructor attempted an SSL_CTX_free operation, leading to a crash).

We also have the legitimate leak problem from use cases like application a loads libcrypto.so via dlopen() and then removes it via dlclose(), and repeats the process in the same process lifetime.  In this situation all globally allocated memory becomes a legitimate leak after the first iteration, as it is no longer reachable.  I know thats an oddball case, but it does happen.

Comment (bob-beck) [reply doc/man7/ossl-guide-migration.pod:67]: How about this for making OPENSSL_cleanup a no-op - sort of.[

https://github.com/openssl/openssl/pull/29494

(needs doc changes if we go with it)

Comment (davidben) [reply doc/man7/ossl-guide-migration.pod:67]: Ah yeah, `dlclose` is fun. My unhelpful answer is that `dlclose` is pronounced "make my process crash in hard-to-debug ways" and so leaks are perfectly fine. :-) But you all may not be willing to just say "don't use this, it's obviously insane".

Fun fact: one of the first OpenSSL debugging adventures of my career was a crash around a OpenSSL ENGINE, where OpenSSL would try to `dlclose` the engine shared object, but the engine registered `ex_data`, because that's how engines registered `free` hooks on data they stored with custom `RSA` objects. There was no way to do this safely at all, because you could not unregister an `ex_data` index.

> And yes, in the case where application A uses libs x and y, and only one participates in the ref counting, we are at risk from crashing in a use after free scenario, but thats no worse than the situation we're in now (see https://github.com/openssl/openssl/issues/25294 in which at atexit cleanup happened before a library destructor attempted an SSL_CTX_free operation, leading to a crash).

Yeah, trying to do complicated logic on shutdown never ends well. But this situation seems worse. Program/library destructors are rare and calling into OpenSSL in then, while it happens, is also rare. Likewise, while some libraries _are_ foolish enough to try to call `OPENSSL_cleanup` today, this is a always bug because a library cannot know what else is going on in the address space.

With this API, libraries will now feel emboldened to participate in the refcount management, particularly if the documentation wrongly describes the API as necessary to avoid memory leaks. But once one library does this, the situation is viral and _all_ libraries have to defensively bump the refcount to avoid problems. (To adapt your example, imagine if the well-meaning but confused library is _itself_ `dlclose`d but OpenSSL itself is never `dlclose`d because there are other users.)

Comment (bob-beck) [reply doc/man7/ossl-guide-migration.pod:67]: > Ah yeah, `dlclose` is fun. My unhelpful answer is that `dlclose` is pronounced "make my process crash in hard-to-debug ways" and so leaks are perfectly fine. :-) But you all may not be willing to just say "don't use this, it's obviously insane".

I too support this view.

Comment (bob-beck) [reply doc/man7/ossl-guide-migration.pod:67]: And I don't know if the right word is "feel emboldend" or "be forced into it because of complainers running valgrind" :) 

If this were an API that normal folks would never need to use or be forced to use, I'd be less concerned.

**Comment by beldmit** [comment]
I am not so sure about any automation in complex cases (multiple load/unload of libcrypto, for example). Probably it's worth running some tests against PKCS#11 provider with unpatched softhsm as backend.

**Comment by nhorman** [comment]
@beldmit I don't suppose you have such a test written/handy to share do you?  I'd like to avoid re-inventing the wheel if possible.

Note also, I requested in https://github.com/openssl/openssl/issues/25294 that you rerun that test with this PR to see if it solves your problem there.  Thank you!

**Comment by beldmit** [comment]
> Note also, I requested in #25294 that you rerun that test with this PR to see if it solves your problem there. Thank you!

It's still on my plate

**Comment by nhorman** [comment]
@beldmit Just as a note, I've attempted to setup openssl+the pkcs11 provider + softHSM2, and it (softHSM2) is having lots of really weird issues running with openssl 4.0 (including some unrelated issues to their locking destructors).  this could take a while.

**Comment by nhorman** [comment]
Ok, I think I've managed to sort out the locking issues with SoftHSM2.  turns out that SoftHSM2 requires that the default provider be loaded, but just assumes that it is, and without it loaded (which I hadn't done), We get into a recursive path that causes deadlock in SoftHSM2 (I've file this [bug for it](https://github.com/softhsm/SoftHSMv2/issues/834)

Unfortunately, we still crash with softhsm2 with destructors in place, in the same way that https://github.com/openssl/openssl/issues/22508.

As a refresher, in that issue we discovered that, in C++ (which is what softHSM2 is written in), global class destructors are implemented in such a way that the glibc atexit() mechanism is used to tear those objects down :(.

Because c++ uses an atexit path to call those destructors, any and all references held to the softhsm library are ignored (i.e. the fact that the library is still loaded via libp11's call to dlopen is irrlevant, the destructors still run on return from the application main routine).

Whats worse is that, atexit handlers run before any gcc destructors run, meaning that we attempt to clean openssl after the softhsm objects are deleted, while we still hold references to that data via openssl->p11 provider->softhsm2.  So when we go to clean up the providers via OPENSSL_cleanup, the p11 provider attempts to derefence freed data, and we get a crash.

We've opened an [issue](https://github.com/softhsm/SoftHSMv2/issues/729) and a [PR](https://github.com/softhsm/SoftHSMv2/pull/742) for this, but they've not been acted on by the SoftHSM maintainers for over a year.

given that this is the same crash that was observed in #22508 , I think the only way we can address this outside of SoftHSM2 is:
1) Use an explicit cleanup mechanism, so that the openssl library cleans up prior to the exit from main (ensuring that the global C++ objects remain valid until the application is done using libcrypto.  That is this PR's solution without the last commit introducing constructors/destructors.

2) Don't ever do any cleanup (which will result in reachable memory not being freed, and possibly some legitimate non-reachable memory, based on what I see in how the p11 provider is written (expecting the provider teardown routine to be called)

3) recognize that c++'s usage of atexit to do global destructor work is no safer than openssl's usage was, and wait for the maintainers there to finish working through https://github.com/softhsm/SoftHSMv2/pull/742

At this point, I'm not sure what the right answer is.

**Comment by nhorman** [comment]
FWIW, I backed out commit 1f7e12f4d78bc44af9ef3d6021fdc1c602b6ba83 from this PR for testing purposes, and, using manual library refcounting, I can make openssl + pkcs11 provider + softHSM2 work without crashing.  By adding add_user and cleanup_ex() calls to The constructor and destructor of the SoftHSM class, I can ensure that openssl gets torn down prior to any atexit() handlers running for global C++ objects in SoftHSM.

I don't think we should have to do this (the global class constructors in SoftHSM are really a problem they need to fix), but at least manual reference counting provides the flexibility to work around it.

**Review by Sashan — CHANGES_REQUESTED** [review]
the PR adds empty lines to files:
   - `fuzz/quic-client.c`
   - `fuzz/quic-rcidm.c`
   - `fuzz/quic-server.c`
   - `fuzz/quic-srtm.c`
   - `fuzz/server.c`

if it is not intentional, then revert those changes to de-clutter the PR. thanks.

File: Configurations/unix-Makefile.tmpl [hunk 3]
```diff
@@ -1779,6 +1779,9 @@ EOF
       die "More than one exported symbol map" if scalar @defs > 1;
 
       my $full = platform->sharedlib($args{lib});
+      my $dirfull = dirname($full);
```
Comment (Sashan) [inline Configurations/unix-Makefile.tmpl:1782]: is change to unix-Makefile.tmpl part of this PR? or is it meant for other issue? I don't see how it is related to library reference counting.

File: Configurations/windows-makefile.tmpl [hunk 4]
```diff
@@ -918,6 +918,7 @@ EOF
      my $deps = join(" ", @objs, @ress, @defs, @deps);
      my $import = platform->sharedlib_import($lib);
      my $dll =  platform->sharedlib($lib);
+     my $dlldir = dirname($dll);
```
Comment (Sashan) [inline Configurations/windows-makefile.tmpl:921]: shall this change be part of this PR? I'm failing to see how it is related to atexit/library reference counting.

File: crypto/init.c [hunk 5]
```diff
@@ -212,8 +228,75 @@ DEFINE_RUN_ONCE_STATIC(ossl_init_async)
...
+         * do the decrement and then load the result to guarantee post decrement evaluation
+         */
+        tsan_decr(&library_users);
+        count = tsan_load(&library_users);
+        OPENSSL_assert(count >= 0);
+        if (count > 0) {
+            return;
+        }
+    } else {
+        /*
```
Comment (Sashan) [inline crypto/init.c:284]: perhaps I'm missing something. The PR introduces library constructor, where we always call `OPENSSL_add_library_user()` where we always set library_refcount_used to one. This constructor should be called by runtime linker when before `dlopen()` returns (explicit linking), or before `main()` is entered (implicit linking to satisfy DT_NEEDED). In either case the point is the `used` is never zero here so we always return.

So it seems to me the branch here can be simplified to:
```
if (legacy_cleanup == 1)
    return;
``` 
The only reason why the extra logic around `library_refcount_used` is to workaround some edgecases when libcrypto is linked statically in that case the constructor might not run. I think it could be issue with OpenBSD runtime linker. But macos/windows/linux do call library constructors even for statically linked binaries.

I'm just scratching my head about this why it is here.

File: crypto/init.c [hunk 6]
```diff
@@ -37,6 +38,21 @@
 #include "s390x_arch.h"
 #endif
 
+#if !defined(OPENSSL_SYS_WIN32)
+__attribute__((constructor)) static void library_constructor(void)
```
Comment (Sashan) [inline crypto/init.c:42]: I'm not sure if `static` is right here. the symbol `library_constructor` should be visible so runtime linker is able to find it and call it. perhaps __attribute__((constructor)) has precedence and does some magic.

File: crypto/init.c [hunk 7]
```diff
@@ -327,6 +410,21 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
     uint64_t tmp;
     int aloaddone = 0;
 
+#if !defined(OPENSSL_SYS_WIN32)
```
Comment (Sashan) [inline crypto/init.c:413]: I'm not entirely convinced this will solve all issues with statically linked libcrypto. the goal here is not to just include the constructor/destructor symbol but to also make sure the constructor gets called before `main()` is entered and destrcutor is called after `main()` is left. We are still at mercy of runtime environement.

File: Configurations/unix-Makefile.tmpl (hunk 3 above)
Comment (nhorman) [reply Configurations/unix-Makefile.tmpl:1782]: Its for the test that got added to test library reference counting.  The test creates two .so objects for the test to load, each of which exercises the proposed refcounting api.  But in building those objects I found that, when creating the link from lib<dso>.so.4 to lib<dso>.so, the pathing was incorrect (i.e. test/lib<dso>.so -> test/lib<dso>.so.4 rather than test/lib<dso>.so -> lib<dso>.so) so this corrects that.  In short, our build system assumes that anything built to the LIBS target is assumed to be in the root directory of the build.

File: Configurations/windows-makefile.tmpl (hunk 4 above)
Comment (nhorman) [reply Configurations/windows-makefile.tmpl:921]: same reasoning as above.

File: crypto/init.c (hunk 5 above)
Comment (nhorman) [reply crypto/init.c:284]: Note: I added the constructor here in response to @bob-beck and @davidben comments [above](https://github.com/openssl/openssl/pull/29466#discussion_r2643596167).  I'm personally not convinced that we should use  them at all (see my comments [here](https://github.com/openssl/openssl/pull/29466#issuecomment-3694029533)).  I was just trying to see if we could make the constructors and destructors work.

> perhaps I'm missing something. The PR introduces library constructor, where we always call OPENSSL_add_library_user() where we always set library_refcount_used to one. This constructor should be called by runtime linker when before dlopen() returns (explicit linking), or before main() is entered (implicit linking to satisfy DT_NEEDED). In either case the point is the used is never zero here so we always return.

You are correct, with functioning constructors/destructors in place, the used value will never be zero, and so OPENSSL_cleanup() would always be a no-op.  However, not all platforms have support for those attributes (i.e. c99 on nonstop and such), so I left the used mechanism in place for those cases.

File: crypto/init.c (hunk 6 above)
Comment (nhorman) [reply crypto/init.c:42]: Visibility doesn't dictate weather or not this symbol gets included in an output object.

In shared objects (dso's), the addition of the constructor and destructor attribute causes gcc/clang to include the symbol universally, where it will be introspected by glibc at library load and executed, weather or not its visible in the symbol export list.

In static objects (.a files), even if you make a symbol public, the linking algorithm causes the symbol to get dropped, since no external code references the function (which is both counter-intuitive, and infuriating), and so, even if you make the symbol public, the constructor/destructor never get called.  To ensure we keep the symbol, we have to internally reference the symbol (see the forced reference I added in OPENSLL_init_crypto() below).

File: crypto/init.c (hunk 7 above)
Comment (nhorman) [reply crypto/init.c:413]: You're right, windows has no notion at all about constructors/destructors in the clang/gcc sense, so we have to call these functions elsewhere (specifcially in the DllMain routines that are built specfically for windows), hence removing these functions for the windows platform here.

**Comment by nhorman** [comment]
closing in favor of https://github.com/openssl/project/issues/1799
